custom/ewsghana/handlers/undo.py

```diff
@@ -7,21 +9,48 @@ class UndoHandler(KeywordHandler):
     def help(self):
         return self.handle()
 
+    def get_user_latest_report(self, user):
+        xform = XFormInstance.view(
+            'reports_forms/all_forms',
+            startkey=['submission xmlns user', self.domain, COMMTRACK_REPORT_XMLNS, user.get_id, {}],
+            endkey=['submission xmlns user', self.domain, COMMTRACK_REPORT_XMLNS, user.get_id],
+            reduce=False,
+            include_docs=True,
+            descending=True
+        ).first()
+        if not xform:
+            return
+        return StockReport.objects.filter(form_id=xform.get_id)
+
+    def restore_message_from_transactions(self, transactions):
+        product_dict = defaultdict(lambda: {'soh': None, 'receipts': 0})
+
+        for tx in transactions:
+            if tx.type == 'receipts':
+                product_dict[tx.sql_product.code]['receipts'] = int(tx.quantity)
+            elif tx.type == 'stockonhand':
+                product_dict[tx.sql_product.code]['soh'] = int(tx.stock_on_hand)
+
+        return 'soh' + ' ' + ' '.join(
+            [
+                '{} {}.{}'.format(code, product_dict[code]['soh'], product_dict[code]['receipts'])
+                for code in sorted(product_dict.keys())
+            ]
+        )
+
     def handle(self):
-        all_prs = StockTransaction.objects.order_by('-report__date')
-        if not all_prs:
-            return self.respond(PRODUCTS_NOT_SUBMITTED)
-
-        sp = all_prs[0].case_id
-        sts = StockTransaction.objects.filter(product_id=all_prs[0].product_id)
-
-        for st in sts:
-            stock_transaction = StockTransaction.objects.get(case_id=sp, product_id=st.product_id)
-            stock_transaction.quantity = st.quantity
-            stock_transaction.save()
-            st.delete()
-
-        self.respond(
-            "Success. Your previous report has been removed. It was: %(report)s",
-            report=sp)
+        last_reports = self.get_user_latest_report(self.user)
+
+        if not last_reports:
+            self.respond('You have not submitted any product reports yet.')
+            return True
+
+        transactions = []
+        for report in last_reports:
+            transactions.extend(report.stocktransaction_set.all())
+
+        message = self.restore_message_from_transactions(transactions)
+        for last_report in last_reports:
+            last_report.delete()
```

Comment:
as mentioned in the fogbugz ticket - I think we should handle this by archiving the form which will remove it from future rebuilds and also [delete the same data](https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/commtrack/models.py#L774)